File: arvores/java/bst/src/bst/BinarySearchTreeSet.java
package bst;

import java.util.ArrayList;
import java.util.Collection;
import java.util.List;

public class BinarySearchTreeSet<K extends Comparable<K>> {

    private int size;
    private Node root;

    public BinarySearchTreeSet() {
        size = 0;
        root = new Node(null, null);
    }

    public BinarySearchTreeSet(Collection<K> c) {
        this();
        addAll(c);
    }

    public int size() {
        return size;
    }

    public boolean isEmpty() {
        return size == 0;
    }

    public boolean remove(K key) {    	
        if (key == null) {
            throw new IllegalArgumentException("Key cannot be null");
        }
    	
        Node nodeToRemove = findKeyLocation(root, key);
        
        if (nodeToRemove.isSentinel()) { // not found
            return false; 
        }

        // two children adjustment
        if (!nodeToRemove.left.isSentinel() && !nodeToRemove.right.isSentinel()) {
            Node successor = findMin(nodeToRemove.right);
            nodeToRemove.key = successor.key;
            nodeToRemove = successor; // nodeToRemove now has 0 or 1 child
        }

        Node child = nodeToRemove.left.isSentinel() ? nodeToRemove.right : nodeToRemove.left;
        child.parent = nodeToRemove.parent;
        
        if (nodeToRemove.parent == null) {
            root = child;
        } else if (nodeToRemove == nodeToRemove.parent.left) {
            nodeToRemove.parent.left = child;
        } else {
            nodeToRemove.parent.right = child;
        }
        
        size--;
        return true;
    }

    private Node findMin(Node node) {
        while (!node.left.isSentinel()) {
            node = node.left;
        }
        return node;
    }
    
    public void add(K key) {
    	if (key == null) {
    		throw new IllegalArgumentException("Key cannot be null");
    	}

        if (isEmpty()) {
            root = new Node(key, null);
            root.left = new Node(null, root);
            root.right = new Node(null, root);
            size++;
            return;
        }

        Node node = findKeyLocation(root, key);
        
        if (node.isSentinel()) { // key not found

        	Node parent = node.parent;
        	Node newNode = new Node(key, parent);
            newNode.left = new Node(null, newNode);
            newNode.right = new Node(null, newNode);

        	if (node == parent.left) {
                parent.left = newNode;
            } else if (node == parent.right) {
                parent.right = newNode;
            }
        	
            size++;
        }
    }

    private Node findKeyLocation(Node node, K key) {
        while (!node.isSentinel()) {
            int cmp = key.compareTo(node.key);
            if (cmp == 0) {
                return node;
            } else if (cmp < 0) {
            	node = node.left;
            } else {
            	node = node.right;
            }
        }
        return node;
    }

    public void addAll(Collection<K> c) {
        for (K item : c) {
            add(item);
        }
    }

    public boolean contains(K key) {
    	Node node = findKeyLocation(root, key);
        return !node.isSentinel();
    }

    public List<K> keys() {
        List<K> keysList = new ArrayList<>();
        collectKeys(root, keysList);
        return keysList;
    }

    private void collectKeys(Node node, List<K> keysList) {
        if (!node.isSentinel()) {
        	collectKeys(node.left, keysList);
            keysList.add(node.key);
            collectKeys(node.right, keysList);
        }
    }
    
    public BinarySearchTreeSet<K> union(BinarySearchTreeSet<K> other) {
        BinarySearchTreeSet<K> result = new BinarySearchTreeSet<>();
        for (K key : this.keys()) {
            result.add(key);
        }
        for (K key : other.keys()) {
            result.add(key);
        }
        return result;
    }

    public BinarySearchTreeSet<K> intersection(BinarySearchTreeSet<K> other) {
        BinarySearchTreeSet<K> result = new BinarySearchTreeSet<>();
        for (K key : this.keys()) {
            if (other.contains(key)) {
                result.add(key);
            }
        }
        return result;
    }

    public BinarySearchTreeSet<K> difference(BinarySearchTreeSet<K> other) {
        BinarySearchTreeSet<K> result = new BinarySearchTreeSet<>();
        for (K key : this.keys()) {
            if (!other.contains(key)) {
                result.add(key);
            }
        }
        return result;
    }
    
    @Override
    public String toString() {
        return keys().toString();
    }

    public String toStringFormat() {
    	StringBuilder sb = new StringBuilder();
    	toStringFormat(root, 0, sb);
    	return sb.toString();
    }
    
    private void toStringFormat(Node node, int depth, StringBuilder sb) {
    	if (!node.isSentinel()) {
        	toStringFormat(node.right, depth + 1, sb);    		
        	String space = (depth > 0) ? "  ".repeat(depth - 1) + "--" : "";
        	String parent = (depth > 0) ? node.parent.key.toString() : "";
        	sb.append(space + "(" + node.key + ")" + parent + "\n");
        	toStringFormat(node.left, depth + 1, sb);
    	}
    }
    
    private class Node {
        K key;
        Node left, right, parent;

        Node(K key, Node parent) {
            this.key = key;
            this.left = this.right = null;
            this.parent = parent;
        }
        
        boolean isSentinel() {
        	return key == null;
        }
    }
}

File: arvores/csharp/bst-set/BinarySearchTreeSet.cs
using System;
using System.Collections.Generic;

namespace Bst
{
    public class BinarySearchTreeSet<K> where K : IComparable<K>
    {
        private int size;
        private Node root;

        public BinarySearchTreeSet()
        {
            size = 0;
            root = new Node(default(K), null);
        }

        public BinarySearchTreeSet(IEnumerable<K> collection) : this()
        {
            AddAll(collection);
        }

        public int Size()
        {
            return size;
        }

        public bool IsEmpty()
        {
            return size == 0;
        }

        public bool Remove(K key)
        {
            if (key == null)
            {
                throw new ArgumentException("Key cannot be null");
            }

            Node nodeToRemove = FindKeyLocation(root, key);

            if (nodeToRemove.IsSentinel)
            {
                return false;
            }

            if (nodeToRemove.Left == null || nodeToRemove.Right == null)
            {
                throw new InvalidOperationException("Left or Right null");
            }

            if (!nodeToRemove.Left.IsSentinel && !nodeToRemove.Right.IsSentinel)
            {
                Node successor = FindMin(nodeToRemove.Right);
                nodeToRemove.Key = successor.Key;
                nodeToRemove = successor;
            }

            if (nodeToRemove.Left == null || nodeToRemove.Right == null)
            {
                throw new InvalidOperationException("Left or Right null");
            }

            Node child = nodeToRemove.Left.IsSentinel ? nodeToRemove.Right : nodeToRemove.Left;
            child.Parent = nodeToRemove.Parent;

            if (nodeToRemove.Parent == null)
            {
                root = child;
            }
            else if (nodeToRemove == nodeToRemove.Parent.Left)
            {
                nodeToRemove.Parent.Left = child;
            }
            else
            {
                nodeToRemove.Parent.Right = child;
            }

            size--;
            return true;
        }

        private Node FindMin(Node node)
        {
            if (node.Left == null)
            {
                throw new InvalidOperationException("Left null");
            }
            while (node.Left != null && !node.Left.IsSentinel)
            {
                node = node.Left;
            }
            return node;
        }

        public void Add(K key)
        {
            if (key == null)
            {
                throw new ArgumentException("Key cannot be null");
            }

            if (IsEmpty())
            {
                root = new Node(key, null);
                root.Left = new Node(default(K), root);
                root.Right = new Node(default(K), root);
                size++;
                return;
            }

            Node node = FindKeyLocation(root, key);

            if (node.IsSentinel && node.Parent != null)
            {
                Node parent = node.Parent;
                Node newNode = new Node(key, parent);
                newNode.Left = new Node(default(K), newNode);
                newNode.Right = new Node(default(K), newNode);

                if (node == parent.Left)
                {
                    parent.Left = newNode;
                }
                else if (node == parent.Right)
                {
                    parent.Right = newNode;
                }

                size++;
            }
        }

        private Node FindKeyLocation(Node node, K key)
        {
            while (!node.IsSentinel)
            {
                int cmp = key.CompareTo(node.Key);
                if (cmp == 0)
                {
                    return node;
                }
                else if (cmp < 0 && node.Left != null)
                {
                    node = node.Left;
                }
                else if (cmp > 0 && node.Right != null)
                {
                    node = node.Right;
                }
            }
            return node;
        }

        public void AddAll(IEnumerable<K> collection)
        {
            foreach (var item in collection)
            {
                Add(item);
            }
        }

        public bool Contains(K key)
        {
            Node node = FindKeyLocation(root, key);
            return !node.IsSentinel;
        }

        public List<K> Keys()
        {
            List<K> keysList = new List<K>();
            CollectKeys(root, keysList);
            return keysList;
        }

        private void CollectKeys(Node? node, List<K> keysList)
        {
            if (node != null && !node.IsSentinel)
            {
                CollectKeys(node.Left, keysList);
                if (node.Key != null)
                {
                    keysList.Add(node.Key);
                }
                CollectKeys(node.Right, keysList);
            }
        }

        public BinarySearchTreeSet<K> Union(BinarySearchTreeSet<K> other)
        {
            BinarySearchTreeSet<K> result = new BinarySearchTreeSet<K>();
            foreach (K key in this.Keys())
            {
                result.Add(key);
            }
            foreach (K key in other.Keys())
            {
                if (!result.Contains(key))
                {
                    result.Add(key);
                }
            }
            return result;
        }

        public BinarySearchTreeSet<K> Intersection(BinarySearchTreeSet<K> other)
        {
            BinarySearchTreeSet<K> result = new BinarySearchTreeSet<K>();
            foreach (K key in this.Keys())
            {
                if (other.Contains(key))
                {
                    result.Add(key);
                }
            }
            return result;
        }

        public BinarySearchTreeSet<K> Difference(BinarySearchTreeSet<K> other)
        {
            BinarySearchTreeSet<K> result = new BinarySearchTreeSet<K>();
            foreach (K key in this.Keys())
            {
                if (!other.Contains(key))
                {
                    result.Add(key);
                }
            }
            return result;
        }

        public override string ToString()
        {
            var keys = Keys();
            if (keys == null || keys.Count == 0)
            {
                return "[]";
            }
            return string.Join(", ", keys.Select(k => k?.ToString() ?? "null"));  // trata chaves nulas
        }

        public string ToStringFormat()
        {
            var sb = new System.Text.StringBuilder();
            ToStringFormat(root, 0, sb);
            return sb.ToString();
        }

        private void ToStringFormat(Node node, int depth, System.Text.StringBuilder sb)
        {
            if (node != null && !node.IsSentinel)
            {
                if (node.Right != null)
                {
                    ToStringFormat(node.Right, depth + 1, sb);
                }
                string space = (depth > 0) ? new string(' ', 2 * (depth - 1)) + "--" : "";
                string parent = (depth > 0) ? (node.Parent?.Key?.ToString() ?? "") : "";
                sb.AppendLine($"{space}({node.Key}){parent}");
                if (node.Left != null)
                {
                    ToStringFormat(node.Left, depth + 1, sb);
                }
            }
        }

        private class Node
        {
            public K? Key { get; set; }
            public Node? Left, Right, Parent;

            public Node(K? key, Node? parent)
            {
                Key = key;
                Left = Right = null;
                Parent = parent;
            }

            public bool IsSentinel => Key == null || Key.Equals(default(K));
        }
    }
}
